nomos-services/mix/src/backends/libp2p.rs

```diff
@@ -0,0 +1,224 @@
+use std::{io, time::Duration};
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use libp2p::{
+    core::transport::ListenerId,
+    identity::{ed25519, Keypair},
+    swarm::SwarmEvent,
+    Multiaddr, PeerId, Swarm, SwarmBuilder, TransportError,
+};
+use nomos_libp2p::{secret_key_serde, DialError, DialOpts, Protocol};
+use overwatch_rs::{overwatch::handle::OverwatchHandle, services::state::NoState};
+use rand::seq::IteratorRandom;
+use serde::{Deserialize, Serialize};
+use tokio::{
+    sync::{broadcast, mpsc},
+    task::JoinHandle,
+};
+
+use super::NetworkBackend;
+
+pub struct Libp2pNetworkBackend {
+    #[allow(dead_code)]
+    task: JoinHandle<()>,
+    msgs_tx: mpsc::Sender<Libp2pNetworkBackendMessage>,
+    events_tx: broadcast::Sender<Libp2pNetworkBackendEvent>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Libp2pNetworkBackendSettings {
+    pub listening_address: Multiaddr,
+    #[serde(with = "secret_key_serde", default = "ed25519::SecretKey::generate")]
+    pub node_key: ed25519::SecretKey,
```

Comment:
Is this key used for both encrypting messages as for the peer id generation?

Reply:
Sorry, we haven't clearly defined what keys are needed in the specification. We have to do it ASAP. @madxor Let's define the [node credentials](https://www.notion.so/Nomos-Network-Tier-1-Persistent-Transmission-Module-10b8f96fb65c807cb1e8f92a7f41a771?pvs=4#11c8f96fb65c80f8b754fd18c331514b) clearly together.

In practice, we have two parts that a certain key should be used (except the incentivization part for now):
1. Establishing "secure" connections
2. Sphinx encryption

The `node_key` in this code is for the 1. Since we're using QUIC and libp2p, the `node_key` is used for [TLS 1.3 handshake](https://docs.libp2p.io/concepts/transports/quic/#quic-in-libp2p) which is automatically done by libp2p. During the handshake, the `node_key` is used to generate a self-signed certificate for authentication. During the handshake, an ephemeral symmetric key is shared between peers, and that is used for encrypting traffic. 

For the 2, I'm thinking of using X25519, which is what we had used in the previous implementation. I didn't add it in our codebase yet because we just use a mock instead of Sphinx. But I'll add it soon.

Reply:
Added a comment to explain the role of `node_key`: 3a6de0887c6f116db0d08377cf0af83f7d673562